source/vir/src/early_exit_cf.rs

```diff
@@ -0,0 +1,161 @@
+use crate::ast::{ArmX, CallTarget, Expr, ExprX, Stmt, StmtX, VirErr};
+use air::ast::Span;
+use air::errors::error;
+
+#[derive(Copy, Clone, Debug)]
+enum StatementType {
+    Return,
+    //Break,
+    //Continue,
+}
+
+#[derive(Clone, Debug)]
+struct EarlyExitInst {
+    pub span: Span,
+    pub statement: StatementType,
+}
+
+pub fn assert_no_early_exit_in_inv_block(inv_span: &Span, expr: &Expr) -> Result<(), VirErr> {
+    let v = expr_get_early_exits(expr);
+    if v.len() == 0 {
+        return Ok(());
+    } else {
+        // TODO improve err msg by adding span for the break/return/continue statement
+        return Err(error("invariant block might return early", inv_span)
+            .primary_label(&v[0].span, "would exit here"));
+    }
+}
+
+/// Walk the AST and find all return/break/continue statements that would cause an
+/// 'early exit' from the expression. Does *not* recurse into nested Invariant blocks,
+/// to avoid quadratic behavior, and to avoid doubling up the errors.
+
+fn expr_get_early_exits(expr: &Expr) -> Vec<EarlyExitInst> {
+    let mut v = Vec::new();
+    expr_get_early_exits_rec(expr, false, &mut v);
+    return v;
+}
+
+/// While recursing, we keep track of whether we entered a loop or not; then we can know
+/// if a break/continue would cause an exit at the high-level expr.
+/// (Well, it will be useful when we implement break/continue, anyway.)
+
+fn expr_get_early_exits_rec(expr: &Expr, in_loop: bool, results: &mut Vec<EarlyExitInst>) {
```

Comment:
In a separate PR, I'm introducing a generic expr visitor that should allow to write these without having to manually recurse. If we think it's worthwhile, we can port this to that later.

Reply:
i wanted to make sure that if somebody adds a feature (e.g., Continue) they'll be forced to update the `match` statement.

Reply:
You may be able to do this with the new visitor, but that's an important point that I'll keep in mind. (And to be clear, I'm not opposed to keep this as separate code if it helps, was just looking for opportunities to reduce the number of visitor implementations.)

Reply:
A good balance would be to use the generic expr/stmt visitor infrastructure, and when using this for the early-exits visitor, implement a minimal visit that just skims each case and does nothing:
```
  ExprX::Tuple(..) => {}
  ExprX::Ctor(..) => {}
...
```
This way we catch any new features that get added, but we don't have to manually implement the recursion into the subexpressions for each case.

Reply:
Yep, that’s what I had in mind too!

Reply:
oh, one problem with this though is that we track extra info in the recursion with the `in_loop` variables (this variable is currently unused, but I wrote this with the expectation we'll eventually need to handle `break` and `continue`). So it can't be done with the generic visitor :\

Reply:
The current generic visitor uses client-defined state (`env`) to handle this; the new generic visitor probably should, too.